WordPress/openverse
# Proof of Concept: Silo ESLint configuration into plugin package

## Description


I've mentioned the possibility of this before and because I'm blocked on local environment issues with Python stuff, I thought I would give this a shot. Especially after working through the issues in #2972 (this PR depends on that one), this particular idea is a bit easier.

The concept here is that all our ESLint related dependencies and configuration get moved into the `@openverse/eslint-plugin` package. This means we only have one place to update ESLint and its related dependencies (plugins, parsers, etc). It makes the configuration at the root of the project very concise as it now just references the plugin and enables one of the built-in configs. I've also gone ahead and split the configuration out in a couple different ways, and there are further improvements available here, especially if we move `overrides` around. This allows us to split the ESLint configuration into easier to understand and navigate files, without creating multiple `eslintrc.js` files in our repository root.

I realised something weird was going on in our ESLint configuration. We had the prettier plugin installed but completely disabled (`prettier/prettier` was `off`). I've removed the plugin itself and just added the `eslint-prettier-config` and add the `prettier` ruleset to `extends` so that the appropriate built-in ESLint rules are disabled. This accomplishes the same end result without the weird situation of enabling the formatting plugin and then disabling it to allow prettier to do its formatting and prevent ESLint from fighting it.

## Testing Instructions

Checkout the branch and run `pnpm install`. Then try out `pnpm run eslint`, confirm it works. Make changes that should cause eslint to fail in various areas (like adding semi-colons, moving imports around, etc) and check that `pnpm run eslint` raises the appropriate errors. Try things in `json`, `vue`, and `ts` files to confirm that all the various plugins are working. A good tip is to remove `eslint-disable` comments and see if those are still working (i.e., if removing them and running `pnpm run eslint` causes errors to appear).

Also confirm that editor integrations work.

## Checklist


- [N/A] I added or updated tests for the changes I made (if applicable).
- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
